relay-server/src/services/buffer/envelope_provider/sqlite.rs

```diff
@@ -0,0 +1,626 @@
+
+            for (_, inner_invelopes) in self.envelopes.iter_mut() {
+                for envelope in inner_invelopes.drain(..) {
+                    if envelopes.len() >= self.disk_batch_size {
+                        Self::flush_many(&mut self.envelope_store, envelopes).await;
+                        envelopes = Vec::with_capacity(self.disk_batch_size);
+                    }
+
+                    envelopes.push(envelope);
+                }
+            }
+
+            if !envelopes.is_empty() {
+                Self::flush_many(&mut self.envelope_store, envelopes).await;
+            }
+
+            self.buffered_envelopes_size = 0;
+        });
+
+        true
+    }
+
+    /// Checks if there's capacity in the store for more envelopes.
+    pub fn has_store_capacity(&self) -> bool {
+        (self.envelope_store.usage() as usize) < self.max_disk_size
+    }
+
+    /// Retrieves the total count of envelopes in the store.
+    pub async fn store_total_count(&self) -> u64 {
+        self.envelope_store
+            .total_count()
+            .await
+            .unwrap_or_else(|error| {
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    "failed to get the total count of envelopes for the sqlite envelope store",
+                );
+                // In case we have an error, we default to communicating a total count of 0.
+                0
+            })
+    }
+
+    /// Returns `true` if there are no [`Envelope`]s for the given [`ProjectKeyPair`], false
+    /// otherwise.
+    fn is_empty(&self, project_key_pair: ProjectKeyPair) -> bool {
+        self.envelopes
+            .get(&project_key_pair)
+            .map_or(true, |e| e.is_empty())
+    }
+
+    /// Determines if the number of buffered envelopes is above the spool threshold.
+    fn above_spool_threshold(&self) -> bool {
+        self.buffered_envelopes_size >= self.disk_batch_size as u64
+    }
+
+    /// Spools envelopes to disk, evenly distributing across all project key pairs.
+    ///
+    /// This method attempts to spool `self.disk_batch_size` envelopes to disk, taking an equal
+    /// number from each group of envelopes (represented by different project key pairs) in the
+    /// `self.envelopes` HashMap.
+    ///
+    /// This approach ensures a balanced spooling across all project key pairs, preventing any
+    /// single group from monopolizing the disk storage.
+    async fn spool_to_disk(&mut self) -> Result<(), SqliteEnvelopeProviderError> {
+        let mut envelopes_to_spool = Vec::with_capacity(self.disk_batch_size);
+
+        // We take envelopes from each project key pair and keep track of which pairs have been
+        // written to disk, to update the disk empty state accordingly.
+        let mut written_project_key_pairs = BTreeSet::new();
+        for (&project_key_pair, envelopes) in self.envelopes.iter_mut() {
+            envelopes_to_spool.append(envelopes);
```

Comment:
Could you add a distribution metric here that logs the avg. / min / max / size of in-memory buffers?